presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownSpatialProjections.java

```diff
@@ -0,0 +1,201 @@
+
+        Expression filter = node.getPredicate();
+        List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filter);
+        if (spatialFunctions.isEmpty()) {
+            return Result.empty();
+        }
+
+        for (FunctionCall spatialFunction : spatialFunctions) {
+            List<Expression> arguments = spatialFunction.getArguments();
+            Verify.verify(arguments.size() == 2);
+
+            Expression firstArgument = arguments.get(0);
+            Expression secondArgument = arguments.get(1);
+
+            Set<Symbol> firstSymbols = extractUnique(firstArgument);
+            Set<Symbol> secondSymbols = extractUnique(secondArgument);
+
+            if (firstSymbols.isEmpty() || secondSymbols.isEmpty()) {
+                return Result.empty();
+            }
+
+            Optional<Symbol> newFirstSymbol = newSymbol(context, firstArgument);
+            Optional<Symbol> newSecondSymbol = newSymbol(context, secondArgument);
+
+            Expression newFirstArgument = toExpression(newFirstSymbol, firstArgument);
+            Expression newSecondArgument = toExpression(newSecondSymbol, secondArgument);
+
+            Expression newSpatialFunction = new FunctionCall(spatialFunction.getName(), ImmutableList.of(newFirstArgument, newSecondArgument));
+            Expression newFilter = replaceExpression(filter, ImmutableMap.of(spatialFunction, newSpatialFunction));
+
+            PlanNode leftNode = joinNode.getLeft();
+            PlanNode rightNode = joinNode.getRight();
+
+            List<Symbol> leftSymbols = leftNode.getOutputSymbols();
+            List<Symbol> rightSymbols = rightNode.getOutputSymbols();
+
+            PlanNode newLeftNode;
+            PlanNode newRightNode;
+
+            if (leftSymbols.containsAll(firstSymbols)
+                    && containsNone(leftSymbols, secondSymbols)
+                    && rightSymbols.containsAll(secondSymbols)
+                    && containsNone(rightSymbols, firstSymbols)) {
+                newLeftNode = newFirstSymbol.map(symbol -> addProjection(context, leftNode, symbol, firstArgument)).orElse(leftNode);
+                newRightNode = newSecondSymbol.map(symbol -> addProjection(context, rightNode, symbol, secondArgument)).orElse(rightNode);
+            }
+            else if (leftSymbols.containsAll(secondSymbols)
+                    && containsNone(leftSymbols, firstSymbols)
+                    && rightSymbols.containsAll(firstSymbols)
+                    && containsNone(rightSymbols, secondSymbols)) {
+                newLeftNode = newSecondSymbol.map(symbol -> addProjection(context, leftNode, symbol, secondArgument)).orElse(leftNode);
+                newRightNode = newFirstSymbol.map(symbol -> addProjection(context, rightNode, symbol, firstArgument)).orElse(rightNode);
+            }
+            else {
+                continue;
+            }
+
+            return Result.ofPlanNode(new JoinNode(
+                    node.getId(),
```

Comment:
it looks like you are missing wrapping `FunctionNode` in case where spatial function had companies

Reply:
I don't think I understand. Would you clarify?

Reply:
I am sorry. I wanted to say `FilterNode` (not function). You got a `FilterNode` as input with a predicate, in which you found a spatial function and you simply returned spatial join. What if there were were other conjuncts?

Reply:
The spatial join includes the full filter (all the conjuncts, not just the spatial function). Similar to hash join, spatial join supports additional filter.